Geo mechanics branch

Vahid-Galavi opened the pull request:
**Description**
The first version of Geomechanics application is added.

Please mark the PR with appropriate tags: 
- Geomechanics application

**Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Geomechanics appliction
- User Defined Soil Models (UDSM) / UMAT using dll (windows) and so (linux)
- Updated Lagrangian UP formulation 
- Staged analysis
- K0-procedure
- Integration with Structural Mechanics
- Automatic time stepping
- ...

Commit:
Geomechanics Application is added

Commit:
- bug in U_Pw_condition.cpp
- styling

Commit:
- styling.

Commit:
- styling.

Commit:
- updated Lagrangian is working for normal U-P elements.

Commit:
- Initialize function with const argument

Commit:
- unused variable is removed.

Commit:
- refactoring small_strain_U_Pw_diff_order_element

Commit:
- refactoring and styling.

Commit:
- updated_lagrangian_U_Pw_diff_order_element is added. It can be compiled but not tested yet.

Commit:
- registration of UpdatedLagrangianUPwDiffOrderElement is added to GeoMechanicsApplication.

Commit:
- Cauchy strain vector and tensor are added.

Commit:
damping matrix of element with different order is added

Commit:
- delete space

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

Vahid-Galavi committed:
Update README.md

philbucher commented on applications/GeoMechanicsApplication/CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+    # Strategies
+)
+
+if(${USE_PORO_MPI} MATCHES ON)
```
we have `USE_MPI` for that
```suggestion
if(${USE_MPI} MATCHES ON)
```

philbucher commented on applications/GeoMechanicsApplication/CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+include_directories( ${CMAKE_SOURCE_DIR}/kratos )
+include_directories( ${CMAKE_SOURCE_DIR}/applications/StructuralMechanicsApplication )
+include_directories( ${CMAKE_SOURCE_DIR}/applications/GeoMechanicsApplication )
+
```
I would strongly suggest to use `kratos_add_dependency` for this, e.g. like [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CableNetApplication/CMakeLists.txt#L8)

philbucher commented on applications/GeoMechanicsApplication/CMakeLists.txt:
```diff
@@ -0,0 +1,167 @@
+    # Processes
+
+    # Constitutive Laws
+    ${CMAKE_CURRENT_SOURCE_DIR}/custom_constitutive/bilinear_cohesive_3D_law.cpp
```
if you check the template app in the Core you can see that we have simplified this a lot

philbucher commented on applications/GeoMechanicsApplication/GeoMechanicsApplication.py:
```diff
@@ -0,0 +1,19 @@
+
+# KM._ImportApplication(application, application_name, application_folder, __path__)
+
+
```
```suggestion
```

we try to not add commented (and in this case old) code

philbucher commented on applications/GeoMechanicsApplication/MainKratos_multiple_stages_template.py:
```diff
@@ -0,0 +1,27 @@
+from __future__ import print_function, absolute_import, division #makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
we dropped python2 support so this can be removed
```suggestion
```

Vahid-Galavi committed:
changes needed to be compiled using gcc

philbucher commented on applications/GeoMechanicsApplication/MainKratos_multiple_stages_template.py:
```diff
@@ -0,0 +1,27 @@
+import KratosMultiphysics.ExternalSolversApplication
+# import KratosMultiphysics.TrilinosApplication as TrilinosApplication
+import KratosMultiphysics.StructuralMechanicsApplication
+import KratosMultiphysics.GeoMechanicsApplication
```
```suggestion
```

I strongly recommend only importing what you need, this is how we do it for the main apps

Vahid-Galavi committed:
Merge branch 'GeoMechanicsBranch' of https://github.com/KratosMultiphysics/Kratos into GeoMechanicsBranch

philbucher reviewed (changes requested):
congrats on finishing your application @Vahid-Galavi 

It seems like you are adding a lot of files that we are trying not to add to the main repo, like large examples, to keep the total size of the repo in reasonable bounds
=> for that we have the examples repo, can you please move your examples there?
(note that the PR should also be [squashed](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/merging-a-pull-request#merging-a-pull-request-on-github) instead of merged to not have the large files in the history)

apart from that I added some general comments, I will continue reviewing once you applied them (currently the PR is too large to keep the overview)

Vahid-Galavi committed:
unused variables are solved

Commit:
Modifications suggested by reviewer

Commit:
replacing all signFactor with one constant

Commit:
CalculateLeftHandSide in small strain UP different order elements

Vahid-Galavi committed:
bug in adding geomteric stiffness matrix of updated lagrangian UP elements with different order

Vahid-Galavi committed:
automatic detection of extension for UDSMs/UMATs

Commit:
bug introduced yesterday is solved for Windows applications

Commit:
- unused function is removed from linear_plane_strain_2D_law.cpp
- refactoring small_strain_U_Pw_diff_order_element
- correcting calculation of gradient of shape function in updated_lagrangian_U_Pw_diff_order_element

Commit:
- internal forces are calculated on initial configuration.

Commit:
- commented block is removed.

Commit:
Geo CR linear beam element (3D2N) is added

Commit:
improving beam elements

Commit:
more improvement in linear beam 3D2N

Commit:
- updating problem type

Commit:
- improvements in beam elements

Commit:
- mass matrix should be calculated on initial configuration
- calculation of derivatives on initial configuration in base class

Commit:
refactoring small strain elements in order to use a similar structure in all elements. In addition ULFEM elements are updated

Commit:
refactoring FIC element

Commit:
- Updated Lagrangian FIC elements.

Commit:
updating CMakeLists.txt

Commit:
updating CMakeLists.txt based on the latest structure

Commit:
- updated lagrangian elements are added to problem type

Vahid-Galavi committed:
modifications needed for GCC

Commit:
- tests are added to test directory.

Commit:
a new process to record results during simulation and to write in files

Commit:
- forgot to commit the python script

Commit:
better solver options

Commit:
- updated problem type with "recording result" part

Commit:
initial value should be written in the file

Vahid-Galavi committed:
unsigned int instead of int

Vahid-Galavi committed:
Merge branch 'GeoMechanicsBranch' of https://github.com/KratosMultiphysics/Kratos into GeoMechanicsBranch

RiccardoRossi commented on applications/GeoMechanicsApplication/license.txt:
```diff
@@ -0,0 +1,7 @@
+Kratos - GeoMechanics Application
```
@KratosMultiphysics/technical-committee  this is a blocker :-(

our policy is to admit as part of the main repository only BSD-licensed applications, so my understanding is that we cannot have a privative licensed application in the main repository. To Be Discussed in @KratosMultiphysics/technical-committee

Vahid-Galavi replied:
Hi @RiccardoRossi,

Our ultimate goal is also to use BSD license. However, this needs some investigation to foresee its consequences. We have chosen this license as a temprary solution to allow further developement/use of Geomechanics application. The application is still open-source but limitted to researchers and academia. Therefore, it should be no problem with the users at the moment.

Cheers,

Vahid-Galavi replied:
@RiccardoRossi : The issue with the license is solved. We switched to BSD license.

Vahid-Galavi replied:
I have already adjusted the license. I hope the code can be merged to the master

RiccardoRossi reviewed (changes requested):
blocking because of the licensing issue. Pending discussion in @KratosMultiphysics/technical-committee

philbucher commented on applications/GeoMechanicsApplication/GeoMechanicsApplication.py:
```diff
@@ -0,0 +1,7 @@
+# Application dependent names and paths
+from KratosMultiphysics import _ImportApplication
+from KratosGeoMechanicsApplication import *
```
here you should also import the Apps this app depends on (aka StructuralMechanics)
Check e.g. the [PotentialFlowApp](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/CompressiblePotentialFlowApplication/CompressiblePotentialFlowApplication.py#L6)

this is needed to avoid nasty errors with undefined symbols

Vahid-Galavi replied:
done!

Commit:
- code review.

Commit:
-typo, -use MACRO to check elements

Commit:
unused Python module should be removed

Commit:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Commit:
updating Geomechanics application with latest changes of Kratos.Core

Vahid-Galavi committed:
defining pointer functions in cpp file instead of header file in order to use windows.h only in cpp file

philbucher commented on applications/GeoMechanicsApplication/custom_constitutive/windows.hpp:
```diff
@@ -0,0 +1,8 @@
+#define NOMINMAX
+#include <windows.h>
+#undef NOMINMAX
+#endif
```
sorry for asking again, but did you try removing this and directly including `#include <windows.h>`?

Since the include is now in a cpp I think it would work

Vahid-Galavi replied:
I replaced windows.hpp with windows.h and removed windows.hpp.

Commit:
- KRATOS_CHECK_VARIABLE_KEY are removed.
- windows.hpp is removed.

Commit:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
const functions

Vahid-Galavi committed:
bug in mass matrix of interface elements. adding info functions in geo elements

Vahid-Galavi committed:
- updating Geomechanics problem type.

Vahid-Galavi committed:
checking variable.Key() is not needed any more

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
- updating GeoMechanicsBranch with master - part 1

Vahid-Galavi committed:
- updating GeoMechanicsBranch with master - part 2.

Vahid-Galavi committed:
- updating GeoMechanicsBranch with master

Vahid-Galavi committed:
- updating GeoMechanicsBranch with master

Vahid-Galavi committed:
adding build scripts for Linux

Vahid-Galavi committed:
- Build scripts for Windows are added.

Vahid-Galavi committed:
- styling

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
updating SetValuesOnIntegrationPoints in GeoMechanicsApplication

Vahid-Galavi committed:
first step in implementation of retention curves

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
- updating README.md and build scripts

Vahid-Galavi committed:
- updating problem type

Vahid-Galavi committed:
- adding retention curve classes

Vahid-Galavi committed:
- adding variables of retention models

Vahid-Galavi committed:
- removing unused modules from Main.py

Vahid-Galavi committed:
- LinearSolversApplication instead of ExternalSolversApplication

Vahid-Galavi committed:
- update pdf

Vahid-Galavi committed:
- add retention model to the elements

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
Merge branch 'GeoMechanicsBranch' of https://github.com/KratosMultiphysics/Kratos into GeoMechanicsBranch

Vahid-Galavi committed:
updating tests to be run locally

Vahid-Galavi committed:
- bug in calculation of Total stress in elements with different orders

Vahid-Galavi committed:
- adding "pore pressure cut off" to cut suction in phreatic levels

Vahid-Galavi committed:
Merge branch 'GeoMechanicsBranch' of https://github.com/KratosMultiphysics/Kratos into GeoMechanicsBranch

Vahid-Galavi committed:
- BSD license for GeoMechanics application.

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
Merge branch 'GeoMechanicsBranch' of https://github.com/KratosMultiphysics/Kratos into GeoMechanicsBranch

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
bug in output flux vector of interface elements

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
check if the size of output vector is correct

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
bug in VanGenuchten model

Vahid-Galavi committed:
- kN instead of N in problem type
- retention law in problem type

Vahid-Galavi committed:
- bug in writing outputs of interface elements is solved.

Vahid-Galavi committed:
- KRATOS_CLASS_POINTER_DEFINITION -> KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION in elements and conditions

Vahid-Galavi committed:
- class instead of namespace in utilities
- cleaning utilities

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
improving MaterialParameters.tcl

Vahid-Galavi committed:
- UMAT parameters are defined as Vector

Vahid-Galavi committed:
- bug in script

Vahid-Galavi committed:
- write UMAT_PARAMETERS instead of PARAMETER_

Vahid-Galavi committed:
- do not store material parameters inside umat/udsm
- all PARAMETER_xxx variables are removed

Vahid-Galavi committed:
changes based on code review

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
- code review

Vahid-Galavi committed:
less includes

Vahid-Galavi committed:
updating geomechanics_solver.py

Vahid-Galavi committed:
updating geomechanics solver similar to structural solver

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
refactoring geomechanics solvers

Vahid-Galavi committed:
refactoring python scripts

Vahid-Galavi committed:
Merge remote-tracking branch 'origin/master' into GeoMechanicsBranch

Vahid-Galavi committed:
some improvements in interface element and bilinear cohesive 3D law

Vahid-Galavi committed:
some improvements in processes

Vahid-Galavi committed:
- gap and closure process (just added and not tested)

Vahid-Galavi committed:
merged with master

Vahid-Galavi committed:
- gap/closure interface process

Vahid-Galavi committed:
- styling and cleaning.

Vahid-Galavi committed:
- styling and cleaning

Vahid-Galavi committed:
- considering gap and closure in interface elements

Vahid-Galavi committed:
- write quadratic interface elements as linear.
